webhook/mutation/pod_mutator.go

```diff
@@ -204,264 +224,474 @@ func isWellKnownWorkload(ownerRef metav1.OwnerReference) bool {
+			Name: oneAgentShareVolumeName,
 			VolumeSource: corev1.VolumeSource{
-				Secret: &corev1.SecretVolumeSource{
-					SecretName: dtingestendpoint.SecretEndpointName,
-				},
+				EmptyDir: &corev1.EmptyDirVolumeSource{},
 			},
 		},
+	)
+}
+
+func ensureInjectionConfigVolume(pod *corev1.Pod) {
+	pod.Spec.Volumes = append(pod.Spec.Volumes,
 		corev1.Volume{
-			Name: "data-ingest-enrichment",
+			Name: injectionConfigVolumeName,
 			VolumeSource: corev1.VolumeSource{
-				EmptyDir: &corev1.EmptyDirVolumeSource{},
+				Secret: &corev1.SecretVolumeSource{
+					SecretName: dtwebhook.SecretConfigName,
+				},
 			},
-		})
+		},
+	)
+}
 
-	var sc *corev1.SecurityContext
-	if pod.Spec.Containers[0].SecurityContext != nil {
-		sc = pod.Spec.Containers[0].SecurityContext.DeepCopy()
+func (m *podMutator) getBasicData(pod *corev1.Pod) (
+	technologies string,
+	installPath string,
+	installerURL string,
+	failurePolicy string,
+	image string,
+) {
+	technologies = url.QueryEscape(kubeobjects.GetField(pod.Annotations, dtwebhook.AnnotationTechnologies, "all"))
+	installPath = kubeobjects.GetField(pod.Annotations, dtwebhook.AnnotationInstallPath, dtwebhook.DefaultInstallPath)
+	installerURL = kubeobjects.GetField(pod.Annotations, dtwebhook.AnnotationInstallerUrl, "")
+	failurePolicy = kubeobjects.GetField(pod.Annotations, dtwebhook.AnnotationFailurePolicy, "silent")
+	image = m.image
+	return
+}
+
+func ensureDkVolume(dk dynatracev1beta1.DynaKube) (corev1.VolumeSource, string) {
+	dkVol := corev1.VolumeSource{}
+	mode := ""
+	if dk.NeedsCSIDriver() {
+		dkVol.CSI = &corev1.CSIVolumeSource{
+			Driver: dtcsi.DriverName,
+		}
+		mode = provisionedVolumeMode
+	} else {
+		dkVol.EmptyDir = &corev1.EmptyDirVolumeSource{}
+		mode = installerVolumeMode
 	}
+	return dkVol, mode
+}
 
-	fieldEnvVar := func(key string) *corev1.EnvVarSource {
-		return &corev1.EnvVarSource{FieldRef: &corev1.ObjectFieldSelector{FieldPath: key}}
+func (m *podMutator) retrieveWorkload(ctx context.Context, req admission.Request, injectionInfo *InjectionInfo, pod *corev1.Pod) (string, string, *admission.Response) {
+	var rsp admission.Response
+	var workloadName, workloadKind string
+	if injectionInfo.enabled(DataIngest) {
+		var err error
+		workloadName, workloadKind, err = findRootOwnerOfPod(ctx, m.metaClient, pod, req.Namespace)
+		if err != nil {
+			rsp = admission.Errored(http.StatusInternalServerError, err)
+			return "", "", &rsp
+		}
 	}
+	return workloadName, workloadKind, nil
+}
 
-	basePodName := pod.GenerateName
-	if basePodName == "" {
-		basePodName = pod.Name
+func (m *podMutator) ensureReinvocationPolicy(pod *corev1.Pod, dk dynatracev1beta1.DynaKube, injectionInfo *InjectionInfo, dataIngestFields map[string]string, req admission.Request) admission.Response {
+	var needsUpdate = false
+	var installContainer *corev1.Container
+	for i := range pod.Spec.Containers {
+		c := &pod.Spec.Containers[i]
+
+		oaInjected := false
+		if injectionInfo.enabled(OneAgent) {
```

Comment:
could this check (and the di check) be handled via annotation?
if not, please move them to a function, that returns whether injection is missing

Reply:
Info if it's enabled is actually based on pods annotations, see: [pod_mutator.go:67](https://github.com/Dynatrace/dynatrace-operator/pull/386/files#diff-e0bb41fcf412fe6b324780881c3a276f59c7e7530df4705e96e33876ff7e0912R66)

Reply:
the enabled check is fine, but the injected check is a bit arbitrary.
any of the other volumes/env variables required for the injection could be missing

Reply:
This logic is [untouched](https://github.com/Dynatrace/dynatrace-operator/blob/3402f37059fd97e79dc4dc598f777706366f4a28/webhook/mutation/pod_mutator.go#L294). It should be changed in separate PR imho.
Agree check is redundant but code looks cleaner and is ready for further refactoring - to hide `injectable` logic behind interfaces.